packages/@aws-cdk/aws-kinesisfirehose-destinations/test/helpers.test.ts

```diff
@@ -0,0 +1,86 @@
+import '@aws-cdk/assert-internal/jest';
+import * as iam from '@aws-cdk/aws-iam';
+import * as logs from '@aws-cdk/aws-logs';
+import * as cdk from '@aws-cdk/core';
+import { createLoggingOptions } from '../lib/private';
+
+describe('createLoggingOptions', () => {
```

Comment:
I really don't like these tests 😕. I don't like testing module-private implementation details. They are bound to change, and all of these tests will then break.

Why can't we write these tests from the perspective of the customer-facing API of `S3Bucket` Destination? That's what actually calls `createLoggingOptions()`, right?

Reply:
Can you explain why directly testing module-private behaviour is a bad thing?

From my perspective, everything that `createLoggingOptions` does is part of the public API, it's just not called directly by the user. If something changes in the function, then our tests should break! It's a reusable abstraction and it should still be consistent.

I see no great benefit of testing it indirectly through `S3Bucket`; just some extra redirection. There's also something to be said about being unsure of whether the behaviour we're testing is actually from `createLoggingOptions` or something `S3Bucket` puts on top. The way it stands currently, we know exactly what `createLoggingOptions` does and can use it in any destination without having to re-test its specifics in the destination tests.

Reply:
> Can you explain why directly testing module-private behaviour is a bad thing?

Sure.

> From my perspective, everything that createLoggingOptions does is part of the public API, it's just not called directly by the user

That's a huge difference though. If it's part of the public API, test it through the public API.

> If something changes in the function, then our tests should break!

No. This makes unit tests much less useful when doing refactoring.

Imagine I changed `createLoggingOptions()` to take an object instead of positional arguments, without changing its behavior. Now, all of your tests will break, while my tests that only use the public API  of the module will continue passing. You will have to change your tests, which means there's nothing helping you confirm that your refactoring of `createLoggingOptions()` was in fact safe. If you test like I'm suggesting, using the public API of the module, the tests will not change, guaranteeing that the refactoring was safe.

Tests should only break if you changed the _behavior_ of something in the system. Not on every change to the implementation details of the system. Your tests here do not have this property.

If you're still not convinced, I would suggest the following resources:

* https://blog.cleancoder.com/uncle-bob/2017/10/03/TestContravariance.html
* https://kentcdodds.com/blog/testing-implementation-details
* https://medium.com/@dev.cprice/stop-testing-implementation-details-77a3528336af

Reply:
I find the changing the interface `createLoggingOptions` by modifying its parameters to be a spurious example of refactoring. The interface of a function is a contract that it binds itself to, not just an implementation detail. A better example of changing an implementation detail would be moving our very long line `const logGroup = props.LogGroup ?? ...` to a separate function or changing the `id` of the `LogGroup` construct that is created. Neither of those changes would cause the current tests to break.

I certainly see the argument of testing the behaviour of this function from the perspective of the public API as this is one thing that we care about: what is the behaviour the S3 bucket destination when the user sets the `logging` property to `true`. However, the other thing we care about is the re-usability of this function from a developer perspective: what is the behaviour of this function in isolation so it can be called from multiple public locations. I think that this latter point is what's missing in _Clean Coder Blog_; there is no mention of what happens when a class that has been extracted from the public class is discovered to be important enough to be re-used elsewhere.

From a practical perspective, what do we do when we add a new destination that also calls `createLoggingOptions`? Do we duplicate the logging tests in this destination as well? Otherwise we have no guarantees on the logging behaviour of this new destination. Now if we change the behaviour of `createLoggingOptions`, we have multiple tests to change, which doesn't feel right either.

Reply:
> I find the changing the interface `createLoggingOptions` by modifying its parameters to be a spurious example of refactoring. The interface of a function is a contract that it binds itself to, not just an implementation detail. A better example of changing an implementation detail would be moving our very long line `const logGroup = props.LogGroup ?? ...` to a separate function or changing the `id` of the `LogGroup` construct that is created. Neither of those changes would cause the current tests to break.

Or, you could use a technique where none of these refactorings break the tests.

(I'm a little scared by you saying the `LogGroup` `id` is an implementation detail - maybe I need to dig into these tests a little bit more. But that's not relevant to this discussion)

> I certainly see the argument of testing the behaviour of this function from the perspective of the public API as this is one thing that we care about: what is the behaviour the S3 bucket destination when the user sets the `logging` property to `true`. However, the other thing we care about is the re-usability of this function from a developer perspective: what is the behaviour of this function in isolation so it can be called from multiple public locations. I think that this latter point is what's missing in _Clean Coder Blog_; there is no mention of what happens when a class that has been extracted from the public class is discovered to be important enough to be re-used elsewhere.

Nothing happens, because the common class is already covered by tests of the public thing it used to be a part of. So extracting it is a safe refactoring, covered by existing tests.

> From a practical perspective, what do we do when we add a new destination that also calls `createLoggingOptions`? Do we duplicate the logging tests in this destination as well? Otherwise we have no guarantees on the logging behaviour of this new destination. Now if we change the behaviour of `createLoggingOptions`, we have multiple tests to change, which doesn't feel right either.

I think you definitely need to test the logging behavior for each destination separately. Look what you're saying here: "I only have to test one kind of logging behavior, because I know the other ones are implemented the same as the first one is". But you only know this as the author of this module - I, as a customer of the module, have no idea that these things share an implementation. When writing tests, you need to put yourself in the shoes of the customer of this code, not its author. Also, this assumption ("I know the other destination's logging are implemented the same as the first one") is not covered by tests itself, and can stop being true in the future - this is how bugs are introduced.

Imagine 6 months from now, someone else on the team comes into this code, and completely re-structures how the logging code is shared (they decide to use a Construct for that sharing, instead of a helper method). Suddenly, not only will the existing tests break, but you have no coverage on the other destinations, which means this refactoring can cause regressions which the unit tests will not catch.

Now, if what you're implying by your question is "I don't want to write these tests again, because they will be duplicated", that's a different thing. We have techniques for dealing with this duplication, like parametrized tests.

I won't hold the PR on this, but in my opinion you're falling here into a classic trap of testing implementation details. It will not cause you problems today, nor tomorrow - the effects will only be felt some time from now, and possibly not by you. That's why it's such an easy mistake to make. I've done this many times in the past, and I just wanted to share this, so you can avoid making the same mistakes as I did 🙂.